### ISSUE TITLE
Feature/ Extend Resolver signature

### ISSUE THREAD

**Issue by theobat** [morpheusgraphql/morpheus-graphql#258]
Hey, awsome work. How hard would it be to add the resolver's spec signature ? Right now there's only the node's arguments passed to the resolver. I'm particularly interested in passing the query's full node AST (and its children)...
(I just discovered your lib, i'm trying to build somthing like [join-monster](https://github.com/acarl005/join-monster) for haskell, I'll look into your lib's code later)

**Comment by nalchevanidze** [comment]
hi, thank you for getting in touch, i don't really understand.  

can you give me an example?

**Comment by theobat** [comment]
So what is currently in the lib is this as far as resolvers are concerned:
```haskell
resolveDeity :: DeityArgs -> IORes Deity
resolveDeity args = gqlResolver $ askDB (name args) (mythology args)

askDB :: Text -> Maybe Text -> IO (Either String Deity)
askDB = ...
```

Whereas in javascript for instance, we'd have this:
```js
function resolveDeity (source, args, context, resolveInfo) {
  // whatever, args contains DeityArgs structure
}
```

The important bits I'm missing from the haskell side are the context and resolveInfo object. resolveInfo essentially has the query's parsed content. For example if the query was like:

```gql
query deity {
  id
  name
}
```

The resolveInfo object would contain info like:
```js
 { fieldName: 'deity',
   fieldNodes:
    [ { kind: 'Field',
        alias: null,
        name: [Object],
        arguments: [Array],
        directives: [],
        selectionSet: null,
        loc: [Object] } ],
   returnType: Deity,
   parentType: RootQuery,
   path: { prev: undefined, key: 'deity' },
// more things such as the entire GQL schema
}
```

And the context object is simply a global object passed down to each resolver (it's useful to pass around an authorization token).

**Comment by theobat** [comment]
Also considering I shared my end goal: the idea is to attach a typeclass (or something like that) to each graphql object within the graphql schema to allow a query resolver to build an SQL query based on what's been asked on the client

**Comment by nalchevanidze** [comment]
i like you idea. it would be great to have something like that in haskell, but there is some problems too.

technically it is possible

```hs
data Query = Query {
   user:: Context Args -> ResolverMonad User
}
```

where

```hs
data Context args = Context {
   args:: args,
   source:: ValidOperation,
   schema:: GQLTypeLib,
}
```

but:

- internal AST is different from `graphql.js` and is not final version. so i lose freedom internally refactoring library without breaking changes.
- validation and resolving are separated and occurs in different step of execution.
  1. parsing :: requestString -> RawOperation
  2. validation :: (RawOperation , GQLTypeLib) -> ValidOperation
  3. execution :: (ValidOperation, resolvers) -> JSON
- to add every resolver possibility to access `GQLSchema` on every resolving. may couse performance problem.

**Comment by theobat** [comment]
OK thanks for the explanation. I'll look into it and see what's possible. For what it's worth, have look at the other graphql libs (https://github.com/haskell-graphql/graphql-api, https://github.com/dmjio/graphql-meta) before starting this one ?

**Comment by nalchevanidze** [comment]
i looked user interface only on: [graphql-api](https://github.com/haskell-graphql/graphql-api) , wrote simple api with it, but after that and decided to write my own GraphQL library

> I totally understand that, but my idea is necessarily tying things with the AST's types so it's fine if it breaks

okey. let see:

then you should look:

https://github.com/morpheusgraphql/morpheus-graphql/blob/1279e132e693a6a88e2affa4aa5b1b72f97f8920/src/Data/Morpheus/Execution/Server/Resolve.hs#L110

there you can pass schema to Resolvers in this case to: `encodeQuery` for query and `encodeOperation` for `muatation` and `subscription`

module Ecnode derives Resolvers
with `Generics`   and generates from it Morpheus `Value` (Value is almost same with JSON value.)

https://github.com/morpheusgraphql/morpheus-graphql/blob/1279e132e693a6a88e2affa4aa5b1b72f97f8920/src/Data/Morpheus/Execution/Server/Encode.hs#L85

`Encode (a -> b) (ResolveT m value)`

a : arguments 
b: result Type

**Comment by nalchevanidze** [comment]
after all to give user correct introspection you may need to adapt:

https://github.com/morpheusgraphql/morpheus-graphql/blob/1279e132e693a6a88e2affa4aa5b1b72f97f8920/src/Data/Morpheus/Execution/Server/Introspect.hs#L86

**Comment by nalchevanidze** [comment]
i would propose that you open your own fork for it. if you have any question i would be glad to answer you.

**Comment by nalchevanidze** [comment]
one possible way could be to define Custom Resolver

```hs
data Context = Context {
  contextSchema    :: DataTypeLib,
  contextSelection :: Selection,
   ...
}

data ContextRes m a = ContextRes {
  unContextRes  :: Context -> m a
}
```

and then we can use it as resolver

```hs
data User m = User {
   name :: UserArgs -> m Text ,
  ..... other fields
}

sqlGetName :: Context  -> IORes Text
sqlGetName context = ...

userResolver :: UserArgs -> ContextRes IORes (User (ContextRes IORes))
userResolver _userArgs = CotextRes $  \userContext -> User { name, ... other fields }
   where 
      name _nameArgs = ContextRes sqlGetName
      .... other resolvers

rootResolver :: GQLRootResolver IO () () (Query (ContextRes IORes)) () ()
rootResolver = ....
```
so you can enable or disable `Context` on query just with a Type  on a `GQLRootResolver`
 - enable context to query : `Query (ContextRes IORes)`
 - disable context to query : `Query IORes`

with this way we should write less code, and is more flexible :) 

can be also used with template Haskell generated resolvers

**Comment by theobat** [comment]
Thanks for all the directions, I'll look into it.

**Comment by theobat** [comment]
So I studied the internals of the lib a little bit, I have trouble understanding why fullSchema and parse & validate request are happening in a

**Comment by theobat** [comment]
So I studied the internals of the lib a little bit, there are a few things I don't understand yet:

- how do you model in the schema the presence of an argument in a nested object ? (e.g. `query test { myObj { id name myNestedObject (filterId: ID) { anotherName } }}`
How is filterId represented here, I've only came accross root (mutation/query/subscription level argument) in the examples

- why is this:
https://github.com/morpheusgraphql/morpheus-graphql/blob/1279e132e693a6a88e2affa4aa5b1b72f97f8920/src/Data/Morpheus/Execution/Server/Resolve.hs#L104-L120
Not isolated in a function ? (I mean is there a reason or is it "just because")
- Your last proposition still requires to change `execOperator` right ? Cause the context and the selection sets still have to be passed ...

**Comment by nalchevanidze** [comment]
anyway. if you think something does not seems to be okey, refactor it. or if you have some questions, i would be glad to answer you either hier or in `slack`

**Comment by theobat** [comment]
Right I'm experimenting/trying things so far :)

**Comment by nalchevanidze** [comment]
hi, i am refactoring and i think it would be soon ready for exposing internal types that you may need

**Comment by theobat** [comment]
Ok, let me know when you think the "Data.Morpheus.Types.Internal.AST.Selection" is "ready-ish" I'd like to add haddock comments first (as a separate PR)

**Comment by nalchevanidze** [comment]
Hi. I think that selection is close to final version. From now I think There will only litle changes

**Comment by nalchevanidze** [comment]
after all we don't need to extend resolver signature at all. look at PR #372

**Comment by theobat** [comment]
Sounds good to me ! Thanks !

**Comment by nalchevanidze** [comment]
done
